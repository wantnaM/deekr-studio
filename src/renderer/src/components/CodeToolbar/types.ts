/**
 * 代码块工具基本信息
 */
export interface CodeToolSpec {
  id: string
  type: 'core' | 'quick'
  order: number
}

/**
 * 代码块工具定义接口
 * @param id 唯一标识符
 * @param type 工具类型
 * @param icon 按钮图标
 * @param tooltip 提示文本
 * @param condition 显示条件
 * @param onClick 点击动作
 * @param order 显示顺序，越小越靠右
 */
export interface CodeTool extends CodeToolSpec {
  icon: React.ReactNode
  tooltip: string
  visible?: () => boolean
  onClick: () => void
}
