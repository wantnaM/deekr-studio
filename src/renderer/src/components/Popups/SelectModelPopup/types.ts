import { Model } from '@renderer/types'
import { ReactNode } from 'react'

// 列表项类型，组名也作为列表项
export type ListItemType = 'group' | 'model'

// 滚动触发来源类型
export type ScrollTrigger = 'initial' | 'search' | 'keyboard' | 'none'

// 扁平化列表项接口
export interface FlatListItem {
  key: string
  type: ListItemType
  icon?: ReactNode
  name: ReactNode
  tags?: ReactNode
  model?: Model
  isPinned?: boolean
  isSelected?: boolean
}

// 滚动和焦点相关的状态类型
export interface ScrollState {
  focusedItemKey: string
  scrollTrigger: ScrollTrigger
  lastScrollOffset: number
  stickyGroup: FlatListItem | null
  isMouseOver: boolean
}

// 滚动和焦点相关的 action 类型
export type ScrollAction =
  | { type: 'SET_FOCUSED_ITEM_KEY'; payload: string }
  | { type: 'SET_SCROLL_TRIGGER'; payload: ScrollTrigger }
  | { type: 'SET_LAST_SCROLL_OFFSET'; payload: number }
  | { type: 'SET_STICKY_GROUP'; payload: FlatListItem | null }
  | { type: 'SET_IS_MOUSE_OVER'; payload: boolean }
  | { type: 'FOCUS_NEXT_ITEM'; payload: { modelItems: FlatListItem[]; step: number } }
  | { type: 'FOCUS_PAGE'; payload: { modelItems: FlatListItem[]; currentIndex: number; step: number } }
  | { type: 'SEARCH_CHANGED'; payload: { searchText: string } }
  | { type: 'FOCUS_ON_LIST_CHANGE'; payload: { modelItems: FlatListItem[] } }
