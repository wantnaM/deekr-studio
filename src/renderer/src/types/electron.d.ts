interface ObsidianAPI {
  getVaults: () => Promise<Array<{ path: string; name: string }>>
  getFiles: (vaultName: string) => Promise<Array<{ path: string; type: 'folder' | 'markdown'; name: string }>>
  getFolders: (vaultName: string) => Promise<Array<{ path: string; type: 'folder' | 'markdown'; name: string }>>
}

interface Window {
  obsidian: ObsidianAPI
}
