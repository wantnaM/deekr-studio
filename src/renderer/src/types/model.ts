import { z } from 'zod'

export const InputType = z.enum(['text', 'image', 'audio', 'video', 'document'])
export type InputType = z.infer<typeof InputType>

export const OutputType = z.enum(['text', 'image', 'audio', 'video', 'vector'])
export type OutputType = z.infer<typeof OutputType>

export const OutputMode = z.enum(['sync', 'streaming'])
export type OutputMode = z.infer<typeof OutputMode>

export const ModelCapability = z.enum([
  'audioGeneration',
  'cache',
  'codeExecution',
  'embedding',
  'fineTuning',
  'imageGeneration',
  'OCR',
  'realTime',
  'rerank',
  'reasoning',
  'streaming',
  'structuredOutput',
  'textGeneration',
  'translation',
  'transcription',
  'toolUse',
  'videoGeneration',
  'webSearch'
])
export type ModelCapability = z.infer<typeof ModelCapability>

export const ModelSchema = z
  .object({
    id: z.string(),
    modelId: z.string(),
    providerId: z.string(),
    name: z.string(),
    group: z.string(),
    description: z.string().optional(),
    owned_by: z.string().optional(),

    supportedInputs: z.array(InputType),
    supportedOutputs: z.array(OutputType),
    supportedOutputModes: z.array(OutputMode),

    limits: z
      .object({
        inputTokenLimit: z.number().optional(),
        outputTokenLimit: z.number().optional(),
        contextWindow: z.number().optional()
      })
      .optional(),

    price: z
      .object({
        inputTokenPrice: z.number().optional(),
        outputTokenPrice: z.number().optional()
      })
      .optional(),

    capabilities: z.array(ModelCapability)
  })
  .refine(
    (data) => {
      // 如果模型支持streaming，则必须支持streamingOutputMode
      if (data.capabilities.includes('streaming') && !data.supportedOutputModes.includes('streaming')) {
        return false
      }

      // 如果模型有OCR能力，则必须支持图像输入类型或者文件输入类型
      if (
        data.capabilities.includes('OCR') &&
        !data.supportedInputs.includes('image') &&
        !data.supportedInputs.includes('document')
      ) {
        return false
      }

      // 如果模型有图像生成能力，则必须支持图像输出
      if (data.capabilities.includes('imageGeneration') && !data.supportedOutputs.includes('image')) {
        return false
      }

      // 如果有音频生成能力，则必须支持音频输出类型
      if (data.capabilities.includes('audioGeneration') && !data.supportedOutputs.includes('audio')) {
        return false
      }

      // 如果有音频识别能力，则必须支持音频输入类型
      if (
        (data.capabilities.includes('transcription') || data.capabilities.includes('translation')) &&
        !data.supportedInputs.includes('audio')
      ) {
        return false
      }

      // 如果有视频生成能力，则必须支持视频输出类型
      if (data.capabilities.includes('videoGeneration') && !data.supportedOutputs.includes('video')) {
        return false
      }

      // 如果模型有embedding能力，则必须支持向量输出类型
      if (data.capabilities.includes('embedding') && !data.supportedOutputs.includes('vector')) {
        return false
      }

      // 如果模型有toolUse, Reasoning, streaming, cache, codeExecution, imageGeneration, audioGeneration, videoGeneration， webSearch能力，则必须支持文字的输入
      if (
        (data.capabilities.includes('toolUse') ||
          data.capabilities.includes('reasoning') ||
          data.capabilities.includes('streaming') ||
          data.capabilities.includes('cache') ||
          data.capabilities.includes('codeExecution') ||
          data.capabilities.includes('imageGeneration') ||
          data.capabilities.includes('audioGeneration') ||
          data.capabilities.includes('videoGeneration') ||
          data.capabilities.includes('webSearch')) &&
        !data.supportedInputs.includes('text')
      ) {
        return false
      }

      // 如果模型有toolUse, Reasoning, streaming, cache, codeExecution, OCR, textGeneration, translation, transcription, webSearch, structuredOutput能力，则必须支持文字的输出
      if (
        (data.capabilities.includes('toolUse') ||
          data.capabilities.includes('reasoning') ||
          data.capabilities.includes('streaming') ||
          data.capabilities.includes('cache') ||
          data.capabilities.includes('codeExecution') ||
          data.capabilities.includes('OCR') ||
          data.capabilities.includes('textGeneration') ||
          data.capabilities.includes('translation') ||
          data.capabilities.includes('transcription') ||
          data.capabilities.includes('webSearch') ||
          data.capabilities.includes('structuredOutput')) &&
        !data.supportedOutputs.includes('text')
      ) {
        return false
      }

      return true
    },
    {
      message: 'ModelCard has inconsistent capabilities and supported input/output type'
    }
  )

export type ModelCard = z.infer<typeof ModelSchema>

export function createModelCard(model: ModelCard): ModelCard {
  return ModelSchema.parse(model)
}

export function supportesInputType(model: ModelCard, inputType: InputType) {
  return model.supportedInputs.includes(inputType)
}

export function supportesOutputType(model: ModelCard, outputType: OutputType) {
  return model.supportedOutputs.includes(outputType)
}

export function supportesOutputMode(model: ModelCard, outputMode: OutputMode) {
  return model.supportedOutputModes.includes(outputMode)
}

export function supportesCapability(model: ModelCard, capability: ModelCapability) {
  return model.capabilities.includes(capability)
}

export function isVisionModel(model: ModelCard) {
  return supportesInputType(model, 'image')
}

export function isImageGenerationModel(model: ModelCard) {
  return isVisionModel(model) && supportesCapability(model, 'imageGeneration')
}

export function isAudioModel(model: ModelCard) {
  return supportesInputType(model, 'audio')
}

export function isAudioGenerationModel(model: ModelCard) {
  return supportesCapability(model, 'audioGeneration')
}

export function isVideoModel(model: ModelCard) {
  return supportesInputType(model, 'video')
}

export function isEmbedModel(model: ModelCard) {
  return supportesOutputType(model, 'vector') && supportesCapability(model, 'embedding')
}

export function isTextEmbeddingModel(model: ModelCard) {
  return isEmbedModel(model) && supportesInputType(model, 'text') && model.supportedInputs.length === 1
}

export function isMultiModalEmbeddingModel(model: ModelCard) {
  return isEmbedModel(model) && model.supportedInputs.length > 1
}

export function isRerankModel(model: ModelCard) {
  return supportesCapability(model, 'rerank')
}

export function isReasoningModel(model: ModelCard) {
  return supportesCapability(model, 'reasoning')
}

export function isToolUseModel(model: ModelCard) {
  return supportesCapability(model, 'toolUse')
}

export function isOnlyStreamingModel(model: ModelCard) {
  return (
    supportesCapability(model, 'streaming') &&
    supportesOutputMode(model, 'streaming') &&
    model.supportedOutputModes.length === 1
  )
}
