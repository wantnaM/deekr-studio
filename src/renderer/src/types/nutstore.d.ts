declare namespace Nutstore {
  export interface FileStat {
    path: string
    basename: string
    isDir: boolean
  }

  type MaybePromise<T> = Promise<T> | T

  export interface Fs {
    ls: (path: string) => MaybePromise<FileStat[]>
    mkdirs: (path: string) => Promise<void>
  }
}
