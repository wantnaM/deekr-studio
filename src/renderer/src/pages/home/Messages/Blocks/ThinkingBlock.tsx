import { CheckOutlined } from '@ant-design/icons'
import { useSettings } from '@renderer/hooks/useSettings'
import { MessageBlockStatus, type ThinkingMessageBlock } from '@renderer/types/newMessage'
import { lightbulbVariants } from '@renderer/utils/motionVariants'
import { Collapse, message as antdMessage, Tooltip } from 'antd'
import { Lightbulb } from 'lucide-react'
import { motion } from 'motion/react'
import { memo, useCallback, useEffect, useMemo, useState } from 'react'
import { useTranslation } from 'react-i18next'
import styled from 'styled-components'

import Markdown from '../../Markdown/Markdown'

interface Props {
  block: ThinkingMessageBlock
}

const ThinkingBlock: React.FC<Props> = ({ block }) => {
  const [copied, setCopied] = useState(false)
  const { t } = useTranslation()
  const { messageFont, fontSize, thoughtAutoCollapse } = useSettings()
  const [activeKey, setActiveKey] = useState<'thought' | ''>(thoughtAutoCollapse ? '' : 'thought')

  const isThinking = useMemo(() => block.status === MessageBlockStatus.STREAMING, [block.status])

  useEffect(() => {
    if (!isThinking && thoughtAutoCollapse) {
      setActiveKey('')
    } else {
      setActiveKey('thought')
    }
  }, [isThinking, thoughtAutoCollapse])

  const copyThought = useCallback(() => {
    if (block.content) {
      navigator.clipboard
        .writeText(block.content)
        .then(() => {
          antdMessage.success({ content: t('message.copied'), key: 'copy-message' })
          setCopied(true)
          setTimeout(() => setCopied(false), 2000)
        })
        .catch((error) => {
          console.error('Failed to copy text:', error)
          antdMessage.error({ content: t('message.copy.failed'), key: 'copy-message-error' })
        })
    }
  }, [block.content, t])

  if (!block.content) {
    return null
  }

  return (
    <CollapseContainer
      activeKey={activeKey}
      size="small"
      onChange={() => setActiveKey((key) => (key ? '' : 'thought'))}
      className="message-thought-container"
      expandIconPosition="end"
      items={[
        {
          key: 'thought',
          label: (
            <MessageTitleLabel>
              <motion.span
                style={{ height: '18px' }}
                variants={lightbulbVariants}
                animate={isThinking ? 'active' : 'idle'}
                initial="idle">
                <Lightbulb size={18} />
              </motion.span>
              <ThinkingText>
                <ThinkingTimeSeconds blockThinkingTime={block.thinking_millsec} isThinking={isThinking} />
              </ThinkingText>
              {/* {isThinking && <BarLoader color="#9254de" />} */}
              {!isThinking && (
                <Tooltip title={t('common.copy')} mouseEnterDelay={0.8}>
                  <ActionButton
                    className="message-action-button"
                    onClick={(e) => {
                      e.stopPropagation()
                      copyThought()
                    }}
                    aria-label={t('common.copy')}>
                    {!copied && <i className="iconfont icon-copy"></i>}
                    {copied && <CheckOutlined style={{ color: 'var(--color-primary)' }} />}
                  </ActionButton>
                </Tooltip>
              )}
            </MessageTitleLabel>
          ),
          children: (
            //  FIXME: 临时兼容
            <div
              style={{
                fontFamily: messageFont === 'serif' ? 'var(--font-family-serif)' : 'var(--font-family)',
                fontSize
              }}>
              <Markdown block={block} />
            </div>
          )
        }
      ]}
    />
  )
}

const ThinkingTimeSeconds = memo(
  ({ blockThinkingTime, isThinking }: { blockThinkingTime?: number; isThinking: boolean }) => {
    const { t } = useTranslation()

    const [thinkingTime, setThinkingTime] = useState(blockThinkingTime || 0)

    // FIXME: 这里统计的和请求处统计的有一定误差
    useEffect(() => {
      let timer: NodeJS.Timeout | null = null
      if (isThinking) {
        timer = setInterval(() => {
          setThinkingTime((prev) => prev + 100)
        }, 100)
      } else if (timer) {
        // 立即清除计时器
        clearInterval(timer)
        timer = null
      }

      return () => {
        if (timer) {
          clearInterval(timer)
          timer = null
        }
      }
    }, [isThinking])

    const thinkingTimeSeconds = useMemo(() => (thinkingTime / 1000).toFixed(1), [thinkingTime])

    return t(isThinking ? 'chat.thinking' : 'chat.deeply_thought', {
      seconds: thinkingTimeSeconds
    })
  }
)

const CollapseContainer = styled(Collapse)`
  margin-bottom: 15px;
`

const MessageTitleLabel = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  height: 22px;
  gap: 4px;
`

const ThinkingText = styled.span`
  color: var(--color-text-2);
`

const ActionButton = styled.button`
  background: none;
  border: none;
  color: var(--color-text-2);
  cursor: pointer;
  padding: 4px;
  display: flex;
  align-items: center;
  justify-content: center;
  margin-left: auto;
  opacity: 0.6;
  transition: all 0.3s;

  &:hover {
    opacity: 1;
    color: var(--color-text);
  }

  &:focus-visible {
    outline: 2px solid var(--color-primary);
    outline-offset: 2px;
  }

  .iconfont {
    font-size: 14px;
  }
`

export default memo(ThinkingBlock)
