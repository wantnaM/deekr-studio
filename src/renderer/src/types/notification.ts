export type NotificationType = 'progress' | 'success' | 'error' | 'warning' | 'info' | 'action'
export type NotificationSource = 'assistant' | 'backup' | 'knowledgeEmbed' | 'update'

export interface Notification<T = any> {
  /** 通知唯一标识 */
  id: string
  /** 通知分类 */
  type: NotificationType
  /** 简要标题，用于列表或弹框的主文案 */
  title: string
  /** 详细描述，可包含执行上下文、结果摘要等 */
  message: string
  /** 时间戳，便于排序与去重 */
  timestamp: number
  /** 可选的进度值（0～1），针对长任务反馈 */
  progress?: number
  /** 附加元数据，T 可定制各种业务字段 */
  meta?: T
  /** 点击或操作回调标识，前端可根据此字段触发路由或函数 */
  actionKey?: string
  /** 声音/声音开关标识，结合用户偏好决定是否播放 */
  silent?: boolean
  /** 渠道：系统级（OS 通知）｜应用内（UI 通知） */
  channel?: 'system' | 'in-app'
  /** 点击回调函数，仅在 type 为 'action' 时有效 */
  onClick?: () => void
  /** 通知源 */
  source: NotificationSource
}
