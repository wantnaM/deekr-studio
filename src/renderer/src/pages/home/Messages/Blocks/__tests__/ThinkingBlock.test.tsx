import type { ThinkingMessageBlock } from '@renderer/types/newMessage'
import { MessageBlockStatus, MessageBlockType } from '@renderer/types/newMessage'
import { render, screen } from '@testing-library/react'
import { act } from 'react'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'

import ThinkingBlock from '../ThinkingBlock'

// Mock dependencies
const mockUseSettings = vi.fn()
const mockUseTranslation = vi.fn()

// Mock hooks
vi.mock('@renderer/hooks/useSettings', () => ({
  useSettings: () => mockUseSettings()
}))

vi.mock('react-i18next', () => ({
  useTranslation: () => mockUseTranslation()
}))

// Mock antd components
vi.mock('antd', () => ({
  Collapse: ({ activeKey, onChange, items, className, size, expandIconPosition }: any) => (
    <div
      data-testid="collapse-container"
      className={className}
      data-active-key={activeKey}
      data-size={size}
      data-expand-icon-position={expandIconPosition}>
      {items.map((item: any) => (
        <div key={item.key} data-testid={`collapse-item-${item.key}`}>
          <div data-testid={`collapse-header-${item.key}`} onClick={() => onChange()}>
            {item.label}
          </div>
          {activeKey === item.key && <div data-testid={`collapse-content-${item.key}`}>{item.children}</div>}
        </div>
      ))}
    </div>
  ),
  Tooltip: ({ title, children, mouseEnterDelay }: any) => (
    <div data-testid="tooltip" title={title} data-mouse-enter-delay={mouseEnterDelay}>
      {children}
    </div>
  ),
  message: {
    success: vi.fn(),
    error: vi.fn()
  }
}))

// Mock icons
vi.mock('@ant-design/icons', () => ({
  CheckOutlined: ({ style }: any) => (
    <span data-testid="check-icon" style={style}>
      âœ“
    </span>
  )
}))

vi.mock('lucide-react', () => ({
  Lightbulb: ({ size }: any) => (
    <span data-testid="lightbulb-icon" data-size={size}>
      ðŸ’¡
    </span>
  )
}))

// Mock motion
vi.mock('motion/react', () => ({
  motion: {
    span: ({ children, variants, animate, initial, style }: any) => (
      <span
        data-testid="motion-span"
        data-variants={JSON.stringify(variants)}
        data-animate={animate}
        data-initial={initial}
        style={style}>
        {children}
      </span>
    )
  }
}))

// Mock motion variants
vi.mock('@renderer/utils/motionVariants', () => ({
  lightbulbVariants: {
    active: { rotate: 10, scale: 1.1 },
    idle: { rotate: 0, scale: 1 }
  }
}))

// Mock Markdown component
vi.mock('@renderer/pages/home/Markdown/Markdown', () => ({
  __esModule: true,
  default: ({ block }: any) => (
    <div data-testid="mock-markdown" data-block-id={block.id}>
      Markdown: {block.content}
    </div>
  )
}))

describe('ThinkingBlock', () => {
  beforeEach(async () => {
    vi.useFakeTimers()

    // Default mock implementations
    mockUseSettings.mockReturnValue({
      messageFont: 'sans-serif',
      fontSize: 14,
      thoughtAutoCollapse: false
    })

    mockUseTranslation.mockReturnValue({
      t: (key: string, params?: any) => {
        if (key === 'chat.thinking' && params?.seconds) {
          return `Thinking... ${params.seconds}s`
        }
        if (key === 'chat.deeply_thought' && params?.seconds) {
          return `Thought for ${params.seconds}s`
        }
        if (key === 'message.copied') return 'Copied!'
        if (key === 'message.copy.failed') return 'Copy failed'
        if (key === 'common.copy') return 'Copy'
        return key
      }
    })
  })

  afterEach(() => {
    vi.restoreAllMocks()
    vi.clearAllMocks()
    vi.clearAllTimers()
    vi.useRealTimers()
  })

  // Test data factory functions
  const createThinkingBlock = (overrides: Partial<ThinkingMessageBlock> = {}): ThinkingMessageBlock => ({
    id: 'test-thinking-block-1',
    messageId: 'test-message-1',
    type: MessageBlockType.THINKING,
    status: MessageBlockStatus.SUCCESS,
    createdAt: new Date().toISOString(),
    content: 'I need to think about this carefully...',
    thinking_millsec: 5000,
    ...overrides
  })

  // Helper functions
  const renderThinkingBlock = (block: ThinkingMessageBlock) => {
    return render(<ThinkingBlock block={block} />)
  }

  const getThinkingContent = () => screen.queryByText(/markdown:/i)
  const getCopyButton = () => screen.queryByRole('button', { name: /copy/i })
  const getThinkingTimeText = () => screen.getByText(/thinking|thought/i)

  describe('basic rendering', () => {
    it('should render thinking content when provided', () => {
      const block = createThinkingBlock({ content: 'Deep thoughts about AI' })
      renderThinkingBlock(block)

      // User should see the thinking content
      expect(screen.getByText('Markdown: Deep thoughts about AI')).toBeInTheDocument()
      expect(screen.getByTestId('lightbulb-icon')).toBeInTheDocument()
    })

    it('should not render when content is empty', () => {
      const testCases = ['', undefined]

      testCases.forEach((content) => {
        const block = createThinkingBlock({ content: content as any })
        const { container, unmount } = renderThinkingBlock(block)
        expect(container.firstChild).toBeNull()
        unmount()
      })
    })

    it('should show copy button only when thinking is complete', () => {
      // When thinking (streaming)
      const thinkingBlock = createThinkingBlock({ status: MessageBlockStatus.STREAMING })
      const { rerender } = renderThinkingBlock(thinkingBlock)

      expect(getCopyButton()).not.toBeInTheDocument()

      // When thinking is complete
      const completedBlock = createThinkingBlock({ status: MessageBlockStatus.SUCCESS })
      rerender(<ThinkingBlock block={completedBlock} />)

      expect(getCopyButton()).toBeInTheDocument()
    })

    it('should match snapshot', () => {
      const block = createThinkingBlock()
      const { container } = renderThinkingBlock(block)
      expect(container.firstChild).toMatchSnapshot()
    })
  })

  describe('thinking time display', () => {
    it('should display appropriate time messages based on status', () => {
      // Completed thinking
      const completedBlock = createThinkingBlock({
        thinking_millsec: 3500,
        status: MessageBlockStatus.SUCCESS
      })
      const { unmount } = renderThinkingBlock(completedBlock)

      const timeText = getThinkingTimeText()
      expect(timeText).toHaveTextContent('3.5s')
      expect(timeText).toHaveTextContent('Thought for')
      unmount()

      // Active thinking
      const thinkingBlock = createThinkingBlock({
        thinking_millsec: 1000,
        status: MessageBlockStatus.STREAMING
      })
      renderThinkingBlock(thinkingBlock)

      const activeTimeText = getThinkingTimeText()
      expect(activeTimeText).toHaveTextContent('1.0s')
      expect(activeTimeText).toHaveTextContent('Thinking...')
    })

    it('should update thinking time in real-time when active', () => {
      const block = createThinkingBlock({
        thinking_millsec: 1000,
        status: MessageBlockStatus.STREAMING
      })
      renderThinkingBlock(block)

      // Initial state
      expect(getThinkingTimeText()).toHaveTextContent('1.0s')

      // After time passes
      act(() => {
        vi.advanceTimersByTime(500)
      })

      expect(getThinkingTimeText()).toHaveTextContent('1.5s')
    })

    it('should handle extreme thinking times correctly', () => {
      const testCases = [
        { thinking_millsec: 0, expectedTime: '0.0s' },
        { thinking_millsec: undefined, expectedTime: '0.0s' },
        { thinking_millsec: 86400000, expectedTime: '86400.0s' }, // 1 day
        { thinking_millsec: 259200000, expectedTime: '259200.0s' } // 3 days
      ]

      testCases.forEach(({ thinking_millsec, expectedTime }) => {
        const block = createThinkingBlock({
          thinking_millsec,
          status: MessageBlockStatus.SUCCESS
        })
        const { unmount } = renderThinkingBlock(block)
        expect(getThinkingTimeText()).toHaveTextContent(expectedTime)
        unmount()
      })
    })

    it('should stop timer when thinking status changes to completed', () => {
      const block = createThinkingBlock({
        thinking_millsec: 1000,
        status: MessageBlockStatus.STREAMING
      })
      const { rerender } = renderThinkingBlock(block)

      // Advance timer while thinking
      act(() => {
        vi.advanceTimersByTime(1000)
      })
      expect(getThinkingTimeText()).toHaveTextContent('2.0s')

      // Complete thinking
      const completedBlock = createThinkingBlock({
        thinking_millsec: 1000, // Original time doesn't matter
        status: MessageBlockStatus.SUCCESS
      })
      rerender(<ThinkingBlock block={completedBlock} />)

      // Timer should stop - text should change from "Thinking..." to "Thought for"
      const timeText = getThinkingTimeText()
      expect(timeText).toHaveTextContent('Thought for')
      expect(timeText).toHaveTextContent('2.0s')

      // Further time advancement shouldn't change the display
      act(() => {
        vi.advanceTimersByTime(1000)
      })
      expect(timeText).toHaveTextContent('2.0s')
    })
  })

  describe('collapse behavior', () => {
    it('should respect auto-collapse setting for initial state', () => {
      // Test expanded by default (auto-collapse disabled)
      mockUseSettings.mockReturnValue({
        messageFont: 'sans-serif',
        fontSize: 14,
        thoughtAutoCollapse: false
      })

      const block = createThinkingBlock()
      const { unmount } = renderThinkingBlock(block)

      // Content should be visible when expanded
      expect(getThinkingContent()).toBeInTheDocument()
      unmount()

      // Test collapsed by default (auto-collapse enabled)
      mockUseSettings.mockReturnValue({
        messageFont: 'sans-serif',
        fontSize: 14,
        thoughtAutoCollapse: true
      })

      renderThinkingBlock(block)

      // Content should not be visible when collapsed
      expect(getThinkingContent()).not.toBeInTheDocument()
    })

    it('should auto-collapse when thinking completes if setting enabled', () => {
      mockUseSettings.mockReturnValue({
        messageFont: 'sans-serif',
        fontSize: 14,
        thoughtAutoCollapse: true
      })

      const streamingBlock = createThinkingBlock({ status: MessageBlockStatus.STREAMING })
      const { rerender } = renderThinkingBlock(streamingBlock)

      // Should be expanded while thinking
      expect(getThinkingContent()).toBeInTheDocument()

      // Stop thinking
      const completedBlock = createThinkingBlock({ status: MessageBlockStatus.SUCCESS })
      rerender(<ThinkingBlock block={completedBlock} />)

      // Should be collapsed after thinking completes
      expect(getThinkingContent()).not.toBeInTheDocument()
    })
  })

  describe('font and styling', () => {
    it('should apply font settings to thinking content', () => {
      const testCases = [
        {
          settings: { messageFont: 'serif', fontSize: 16 },
          expectedFont: 'var(--font-family-serif)',
          expectedSize: '16px'
        },
        {
          settings: { messageFont: 'sans-serif', fontSize: 14 },
          expectedFont: 'var(--font-family)',
          expectedSize: '14px'
        }
      ]

      testCases.forEach(({ settings, expectedFont, expectedSize }) => {
        mockUseSettings.mockReturnValue({
          ...settings,
          thoughtAutoCollapse: false
        })

        const block = createThinkingBlock()
        const { unmount } = renderThinkingBlock(block)

        // Find the styled content container
        const contentContainer = screen.getByTestId('collapse-content-thought')
        const styledDiv = contentContainer.querySelector('div')

        expect(styledDiv).toHaveStyle({
          fontFamily: expectedFont,
          fontSize: expectedSize
        })

        unmount()
      })
    })
  })

  describe('integration and edge cases', () => {
    it('should handle content updates correctly', () => {
      const block1 = createThinkingBlock({ content: 'Original thought' })
      const { rerender } = renderThinkingBlock(block1)

      expect(screen.getByText('Markdown: Original thought')).toBeInTheDocument()

      const block2 = createThinkingBlock({ content: 'Updated thought' })
      rerender(<ThinkingBlock block={block2} />)

      expect(screen.getByText('Markdown: Updated thought')).toBeInTheDocument()
      expect(screen.queryByText('Markdown: Original thought')).not.toBeInTheDocument()
    })

    it('should clean up timer on unmount', () => {
      const block = createThinkingBlock({ status: MessageBlockStatus.STREAMING })
      const { unmount } = renderThinkingBlock(block)

      const clearIntervalSpy = vi.spyOn(global, 'clearInterval')
      unmount()

      expect(clearIntervalSpy).toHaveBeenCalled()
    })

    it('should handle rapid status changes gracefully', () => {
      const block = createThinkingBlock({ status: MessageBlockStatus.STREAMING })
      const { rerender } = renderThinkingBlock(block)

      // Rapidly toggle between states
      for (let i = 0; i < 3; i++) {
        rerender(<ThinkingBlock block={createThinkingBlock({ status: MessageBlockStatus.STREAMING })} />)
        rerender(<ThinkingBlock block={createThinkingBlock({ status: MessageBlockStatus.SUCCESS })} />)
      }

      // Should still render correctly
      expect(getThinkingContent()).toBeInTheDocument()
      expect(getCopyButton()).toBeInTheDocument()
    })
  })
})
