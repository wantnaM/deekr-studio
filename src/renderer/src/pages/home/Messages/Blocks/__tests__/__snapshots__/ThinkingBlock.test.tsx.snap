// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThinkingBlock > basic rendering > should match snapshot 1`] = `
.c0 {
  margin-bottom: 15px;
}

.c1 {
  display: flex;
  flex-direction: row;
  align-items: center;
  height: 22px;
  gap: 4px;
}

.c2 {
  color: var(--color-text-2);
}

.c3 {
  background: none;
  border: none;
  color: var(--color-text-2);
  cursor: pointer;
  padding: 4px;
  display: flex;
  align-items: center;
  justify-content: center;
  margin-left: auto;
  opacity: 0.6;
  transition: all 0.3s;
}

.c3:hover {
  opacity: 1;
  color: var(--color-text);
}

.c3:focus-visible {
  outline: 2px solid var(--color-primary);
  outline-offset: 2px;
}

.c3 .iconfont {
  font-size: 14px;
}

<div
  class="c0 message-thought-container"
  data-active-key="thought"
  data-expand-icon-position="end"
  data-size="small"
  data-testid="collapse-container"
>
  <div
    data-testid="collapse-item-thought"
  >
    <div
      data-testid="collapse-header-thought"
    >
      <div
        class="c1"
      >
        <span
          data-animate="idle"
          data-initial="idle"
          data-testid="motion-span"
          data-variants="{"active":{"rotate":10,"scale":1.1},"idle":{"rotate":0,"scale":1}}"
          style="height: 18px;"
        >
          <span
            data-size="18"
            data-testid="lightbulb-icon"
          >
            ðŸ’¡
          </span>
        </span>
        <span
          class="c2"
        >
          Thought for 5.0s
        </span>
        <div
          data-mouse-enter-delay="0.8"
          data-testid="tooltip"
          title="Copy"
        >
          <button
            aria-label="Copy"
            class="c3 message-action-button"
          >
            <i
              class="iconfont icon-copy"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      data-testid="collapse-content-thought"
    >
      <div
        style="font-family: var(--font-family); font-size: 14px;"
      >
        <div
          data-block-id="test-thinking-block-1"
          data-testid="mock-markdown"
        >
          Markdown: 
          I need to think about this carefully...
        </div>
      </div>
    </div>
  </div>
</div>
`;
